### ISSUE TITLE
Express and explicitly stating types

### ISSUE THREAD

**Issue by rb1147x** [haxiomic/dts2hx#76]
I'm starting a new project and this is my first time using the generated express externs.

Looking at the example:

```
var app = Express.call();
var port = 3000;

app.get('/', (req, res, next) -> res.send('Hello World! <a href="/birds">/birds</a>'));
app.listen(port, () -> trace('Example app listening at http://localhost:${port}'));

app.get('/route', (req, res, next) -> res.send('Example route'));
```

I've noticed that types are not explicitly stated (e.g. `(req:Request, res:Response, next:Middleware)` like i've been able to do when using [other handmade externs](https://github.com/clemos/haxe-js-kit)) and I seem to be stuck on figuring out what they should be when I try to add them, probably because the externs are generated with type paramaters and am unsure how to handle those.

Basically, how would I write this code, replacing the `?` with the type:

```
app.get('/', function(req:?, res:?, next:NextFunction) {
    req.send(html);
});

```

While I personally like the explicit style, without stating the types, I'm missing out on type hinting, autocompletion, etc.

Also, looking through the generated externs, it seems `Response.hx`, `Request.hx`, and many other files are generated as typedefs when I would have imagined them to be generated as classes. Are they supposed to be generated as typedefs? Why is that?

Thanks.

**Comment by haxiomic** [comment]
Hey @rb1147x, autocompletion and type hints should be working without type annotations:

![express-type-hint](https://user-images.githubusercontent.com/3742992/102357026-d0ca3a80-3fa5-11eb-8b20-0c2d811d4a85.gif)

However you can manually specify if you'd prefer
```haxe
app.get('/', (req:Request<Any, Any, Any, Any>, res: Response<Any, Float>, next: NextFunction) -> {});
```
The idea with the type parameters here is their type can be often automatically determined by the other parameters

In typescript you can have default type parameter values, so you don't need to set them all, but this isn't available in haxe yet (however there is a [proposal](https://github.com/HaxeFoundation/haxe-evolution/pull/50))

Request and Response are both interfaces in TypeScript, ts interface has a different meaning to haxe interface, in TypeScript and interface is very flexible and maps much more closely to a haxe typedef

**Comment by rb1147x** [comment]
Thanks for the explanations. I do not see the hinting you see. In fact just copying the example code there's already a "could not find suitable overload" error:

![error](https://user-images.githubusercontent.com/17628197/102374249-bc2a7a00-3f86-11eb-9a36-89a50d4491ea.gif)

Adding the annotations you suggested still results in an error:

![e2](https://user-images.githubusercontent.com/17628197/102374883-7621e600-3f87-11eb-8d07-328ff183e17a.PNG)

And the error output:

![e3](https://user-images.githubusercontent.com/17628197/102375280-f34d5b00-3f87-11eb-8583-95140d14f379.PNG)

**Comment by haxiomic** [comment]
hmm, looks like something's askew with the environment somewhere. Here's how I can reproduce my case where it works:

I've tested this with haxe 4.0.5, 4.1.4 and 4.2 nightly all compile without error, which haxe version have you got? (`haxe --version`)

In a new directory:
```
 npm i express @types/express dts2hx
 npx dts2hx express
 ```
Then copy in Server.hx from the example and
```
haxe --library express --main Server.hx --js server.js
node server.js

Server.hx:13: Example app listening at http://localhost:3000
```

**Comment by haxiomic** [comment]
if it compiles on the command line, but vscode is giving errors, that implies something funky with vscode – maybe it's using a different version of haxe to the one available on the command line. Make sure to restart vscode if you change system haxe version

**Comment by rb1147x** [comment]
Okay, I was able to partially figure it out. For reference I'm using Haxe 4.1.4.

I'm not sure why hinting is still coming up empty - vshaxe shows no errors, but no hints without annotations:

![e1](https://user-images.githubusercontent.com/17628197/102382374-0106de80-3f90-11eb-924c-e47435904a8f.PNG)

As far as the type annotations, the modules needs to be from the `express_serve_static_core` package (am I phrasing that correctly?) and once they are added, vshaxe is able to provide hints, completion, etc:

![2](https://user-images.githubusercontent.com/17628197/102382770-6c50b080-3f90-11eb-9981-c6cae84c0f9a.PNG)

![hints](https://user-images.githubusercontent.com/17628197/102383231-f00a9d00-3f90-11eb-901f-7fbc7dfdf4cb.PNG)

Still leaves the unanswered question as to why vshaxe isn't able to provide hinting without the annotations in my environment, but I'm okay with that for now if it'll work with the annotations included. So far this seems to just be a problem with express.

**Comment by haxiomic** [comment]
This might be an IDE edge case we can improve, I have the same behavior when I mouse over `req` and `res`, but when I mouse into the whitespace between them it shows the full function signature with the types for `req` and `res`, so the information is there, it's just getting lost in the IDE display somehow

glad it's working for you now in any case. The `express` vs `express_serve_static_core` is a nuisance I know, surprisingly this is actually how the types are defined in ts, but it's just easier to obscure there so the user doesn't see that they come from the "express static core" module. I've got some thoughts on how to improve this on the roadmap so maybe one day it'll be a little cleaner there

**Comment by haxiomic** [comment]
@rb1147x  Ahhh, little update: I see the behavior you're having now – when I use haxe 4.1.4 I don't get autocomplete on `req` either, but I do with haxe nightly, I guess it was a known issue that's been fixed in the compiler

**Comment by rb1147x** [comment]
Oh excellent! Thanks for the update!
